node/core/av-store/src/lib.rs

```diff
@@ -195,65 +572,234 @@ fn available_data(db: &Arc<dyn KeyValueDB>, candidate_hash: &Hash) -> Option<Sto
+
+
+async fn get_block_number<Context>(
+	ctx: &mut Context,
+	relay_parent: Hash,
+) -> Result<BlockNumber, Error>
+where
+	Context: SubsystemContext<Message=AvailabilityStoreMessage>,
+{
+	let (tx, rx) = oneshot::channel();
+
+	ctx.send_message(AllMessages::ChainApi(ChainApiMessage::BlockNumber(relay_parent, tx))).await?;
+
+	Ok(rx.await??.map(|number| number + 1).unwrap_or_default())
+}
+
+fn store_available_data(
+	subsystem: &mut AvailabilityStoreSubsystem,
 	candidate_hash: &Hash,
 	id: Option<ValidatorIndex>,
 	n_validators: u32,
 	available_data: AvailableData,
-	metrics: &Metrics,
 ) -> Result<(), Error> {
 	let mut tx = DBTransaction::new();
 
+	let block_number = available_data.validation_data.block_number;
+
 	if let Some(index) = id {
-		let chunks = get_chunks(&available_data, n_validators as usize, metrics)?;
-		store_chunk(db, candidate_hash, n_validators, chunks[index as usize].clone())?;
+		let chunks = get_chunks(&available_data, n_validators as usize, &subsystem.metrics)?;
+		store_chunk(
+			subsystem,
+			candidate_hash,
+			n_validators,
+			chunks[index as usize].clone(),
+			block_number,
+		)?;
 	}
 
 	let stored_data = StoredAvailableData {
 		data: available_data,
 		n_validators,
 	};
 
+	let mut pov_pruning = pov_pruning(&subsystem.inner).unwrap_or_default();
+
+	let pruning_record = PoVPruningRecord {
+		candidate_hash: candidate_hash.clone(),
+		block_number,
+		candidate_state: CandidateState::Stored,
+		prune_at: SystemTime::now().duration_since(UNIX_EPOCH)?.as_secs() +
+			subsystem.pruning_config.keep_stored_block_for,
+	};
+
+	let idx = pov_pruning.binary_search(&pruning_record).unwrap_or_else(|x| x);
+
+	pov_pruning.insert(idx, pruning_record);
+	let next_pruning = pov_pruning[0].prune_at;
+
 	tx.put_vec(
 		columns::DATA,
 		available_data_key(&candidate_hash).as_slice(),
 		stored_data.encode(),
 	);
 
-	db.write(tx)?;
+	tx.put_vec(
+		columns::META,
+		&POV_PRUNING_KEY,
+		pov_pruning.encode(),
+	);
+
+	tx.put_vec(
+		columns::META,
+		&NEXT_POV_PRUNING,
+		NextPoVPruning(next_pruning).encode(),
+	);
+
+	subsystem.inner.write(tx)?;
 
 	Ok(())
 }
 
-fn store_chunk(db: &Arc<dyn KeyValueDB>, candidate_hash: &Hash, _n_validators: u32, chunk: ErasureChunk)
-	-> Result<(), Error>
-{
+fn store_chunk(
+	subsystem: &mut AvailabilityStoreSubsystem,
+	candidate_hash: &Hash,
+	_n_validators: u32,
+	chunk: ErasureChunk,
+	block_number: BlockNumber,
+) -> Result<(), Error> {
 	let mut tx = DBTransaction::new();
 
 	let dbkey = erasure_chunk_key(candidate_hash, chunk.index);
 
-	tx.put_vec(columns::DATA, &dbkey, chunk.encode());
-	db.write(tx)?;
+	let mut chunk_pruning = chunk_pruning(&subsystem.inner).unwrap_or_default();
+
+	let pruning_record = ChunkPruningRecord {
+		candidate_hash: candidate_hash.clone(),
+		block_number,
+		candidate_state: CandidateState::Stored,
+		chunk_index: chunk.index,
+		prune_at: SystemTime::now().duration_since(UNIX_EPOCH)?.as_secs() +
+			subsystem.pruning_config.keep_stored_block_for,
```

Comment:
minor issue: should we use checked / saturating add. The concern is that `keep_stored_block_for` comes from the outside of this module and there are no preconditions imposed on it.

Reply:
I think `keep_stored_block_for` is always a constant expect for tests.

Reply:
ah, `PruningConfig` for `new_on_disk` is always default.